pkg/controller/certificates/signer/signer.go

```diff
@@ -89,6 +89,19 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
 }
 
 func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
+	// Special-case handling here in case we are talking to an older apiserver.
+	if csr.Spec.SignerName == nil {
+		return nil
+	}
+	// Only handle CSRs with the specific known signerNames.
```

Comment:
for this, I expected a switch on the resolved (maybe defaulted above) signerName:

```go
switch signerName {
case capi.KubeletServingSignerName:
  // also make sure IsKubeletServingCert passes on the CSR
case capi.KubeAPIServerClientKubeletSignerName:
  // also make sure IsAPIServerClientKubelet passes on the CSR
case capi.KubeAPIServerClientSignerName:
  // also make sure IsAPIServerClient passes on the CSR
case capi.LegacyUnknownSignerName:
  // any additional validation?
default:
  // ignore other signerNames
  return nil
}
```

Reply:
> // also make sure IsAPIServerClient passes on the CSR

This does not currently exist as we don't have a definition/any restrictions on `kube-apiserver-client` (only `kube-apiserver-client-kubelet`).

The subject restrictions mentioned in the KEP are noted as being enforced during *admission* and not during approval/signing (related: @mikedanese's question here: https://github.com/kubernetes/enhancements/pull/1400#discussion_r374399364).

Should we:

1) enforce in all of admission, approval and signing? The additional check seems redundant unless we want to encourage restricting `system:masters` usage across approvers/signers.
2) enforce in approval and signing only?

Reply:
Ah, I see there are also additional restrictions to apply defined in the KEP:

```
Trust distribution: signed certificates must be honored as client-certificates by the kube-apiserver. The CA bundle is not distributed by any other means.
Permitted subjects - no subject restrictions, but approvers and signers may choose not to approve or sign. Certain subjects like cluster-admin level users or groups vary between distributions and installations, but deserve additional scrutiny before approval and signing. An admission plugin is available to restrict system:masters, but it is often not the only cluster-admin subject in a cluster.
Permitted x509 extensions - Non-critical extensions should be dropped.
Permitted key usages - must include []string{"client auth"}. Must not include key usages beyond []string{"digital signature", "key encipherment", "client auth"}
Expiration/cert lifetime - minimum of CSR signer or request. Sanity of the time is the concern of the signer.
CA bit allowed/disallowed - not allowed.
```

Reply:
Yeah, define a helper function that encapsulates the extensions/usages/ca requirements and call it here before signing a CSR that requests that signer name.